framework/src/base/NonlinearSystem.C

```diff
@@ -149,7 +144,25 @@ NonlinearSystem::solve()
   setInitialSolution();
 
   if (_use_finite_differenced_preconditioner)
+  {
+    _transient_sys.nonlinear_solver->fd_residual_object = &_fd_residual_functor;
     setupFiniteDifferencedPreconditioner();
+  }
+
+#ifdef LIBMESH_HAVE_PETSC
+  PetscNonlinearSolver<Real> & solver =
+      static_cast<PetscNonlinearSolver<Real> &>(*_transient_sys.nonlinear_solver);
+  Moose::SolveType moose_solve_type = _fe_problem.solverParams()._type;
+  if (moose_solve_type == Moose::ST_PJFNK || moose_solve_type == Moose::ST_JFNK)
```

Comment:
There is a situation, where code may not work properly.  If we setup a matrix free method through a PETSc command-line option "`-snes_mf_operator 1`", and then the things in MOOSE and Libmesh may not be setup properly.

Reply:
If `-snes_mf_operator 1` is given in the input file, we already warn users that they should not do that (see [here](https://github.com/idaholab/moose/blob/devel/framework/src/utils/PetscSupport.C#L655). If it is specified on the command line, then that would only affect contact.

Reply:
I had thought of this before, but I don't have a good suggestion for fixing it...

Reply:
So @fdkong, per our discussion yesterday, this is something that should be fixable by reordering/refactor the `PetscNonlinearSolver::solve` in libmesh

Reply:
Pushed changes up to Libmesh at https://github.com/libMesh/libmesh/pull/1572/files